File: beakjoon/2667/2667.py
N = int(input())

m = []
visited = [[False for __ in range(N)] for _ in range(N)]

for i in range(N):
    m.append(list(map(int, input())))

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

counter = [0 for _ in range(N*N)]


def dfs(x, y, count, idx):
    global counter

    visited[x][y] = True
    counter[idx] += 1

    for i in range(4):
        nx = x+dx[i]
        ny = y+dy[i]

        if nx >= 0 and nx < N and ny >= 0 and ny < N \
                and m[x][y] == m[nx][ny] and not visited[nx][ny]:
            dfs(nx, ny, count+1, idx)


idx = 0
for i in range(N):
    for j in range(N):
        if not visited[i][j] and m[i][j] != 0:
            dfs(i, j, 0, idx)
            idx += 1

print(idx)
for n in sorted(counter[:idx]):
    print(n)

File: beakjoon/2667/2667.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};
int N, idx = 0;
void dfs(vector<int> *map, vector<bool> *visited, vector<int> &counter, int x, int y, int count, int idx)
{
    visited[x][y] = true;
    counter[idx]++;

    for (int i = 0; i < 4; i++)
    {
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (nx >= 0 && nx < N && ny >= 0 && ny < N && map[x][y] == map[nx][ny] && !visited[nx][ny])
            dfs(map, visited, counter, nx, ny, count + 1, idx);
    }
}
int main()
{

    cin >> N;

    vector<int> map[N];
    vector<bool> visited[N];
    vector<int> counter(N * N, 0);

    for (int i = 0; i < N; i++)
    {
        string tmp;
        cin >> tmp;
        for (int j = 0; j < N; j++)
        {
            map[i].push_back(tmp[j] - '0');
            visited[i].push_back(false);
        }
    }

    for (int i = 0; i < N; i++)
        for (int j = 0; j < N; j++)
            if (map[i][j] != 0 && !visited[i][j])
                dfs(map, visited, counter, i, j, 0, idx++);

    sort(counter.begin(), counter.begin() + idx);
    cout << idx << "\n";
    for (int i = 0; i < idx; i++)
        cout << counter[i] << "\n";

    return 0;
}